File: sdks/rust/test-optimization-sdk/src/test_optimization/mock_tracer.rs
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2025 Datadog, Inc.

//! Mock tracer module for testing and debugging purposes
//!
//! This module provides functionality for creating and managing mock spans,
//! setting tags, error information, and closing spans.
use crate::test_optimization::lib::*;
use crate::test_optimization::utils::*;
use std::collections::HashMap;
use std::ffi::CStr;
use std::time::{Duration, SystemTime};

#[derive(Debug, Clone)]
/// Represents a mock span for testing and debugging purposes
pub struct MockSpan {
    /// The unique identifier for this span
    #[allow(dead_code)]
    pub span_id: u64,
    /// The ID of the trace
    #[allow(dead_code)]
    pub trace_id: u64,
    /// The ID of the parent span, if any
    #[allow(dead_code)]
    pub parent_span_id: u64,
    /// The start time of the span
    #[allow(dead_code)]
    pub start_time: SystemTime,
    /// The end time of the span
    #[allow(dead_code)]
    pub finish_time: SystemTime,
    /// The name of the operation being traced
    #[allow(dead_code)]
    pub operation_name: String,
    /// The tags associated with this span
    #[allow(dead_code)]
    pub string_tags: HashMap<String, String>,
    /// The numeric tags associated with this span
    #[allow(dead_code)]
    pub number_tags: HashMap<String, f64>,
}

#[derive(Debug, Clone)]
/// Represents a mock tracer for testing and debugging purposes
pub struct MockTracer;

impl MockTracer {
    /// Resets the mock tracer to its initial state
    #[allow(dead_code)]
    pub fn reset() -> bool {
        unsafe {
            Bool_to_bool(topt_debug_mock_tracer_reset())
        }
    }

    /// Returns a vector of all finished spans in this mock tracer
    #[allow(dead_code)]
    pub fn get_finished_spans() -> Vec<MockSpan> {
        unsafe {
            // Get the array from the native side.
            let finished_array = topt_debug_mock_tracer_get_finished_spans();
            // Convert the C array into a Vec<MockSpan>.
            let spans = Self::convert_mock_span_array(&finished_array);
            // Free the native array.
            topt_debug_mock_tracer_free_mock_span_array(finished_array);
            // Return
            spans
        }
    }

    /// Returns a vector of all open spans in this mock tracer
    #[allow(dead_code)]
    pub fn get_open_spans() -> Vec<MockSpan> {
        unsafe {
            // Get the array from the native side.
            let open_array = topt_debug_mock_tracer_get_open_spans();
            // Convert the C array into a Vec<MockSpan>.
            let spans = Self::convert_mock_span_array(&open_array);
            // Free the native array.
            topt_debug_mock_tracer_free_mock_span_array(open_array);
            // Return
            spans
        }
    }

    /// Converts a topt_UnixTime to a SystemTime
    fn convert_unix_time(ut: &topt_UnixTime) -> SystemTime {
        SystemTime::UNIX_EPOCH + Duration::new(ut.sec, ut.nsec as u32)
    }

    /// Converts a C KeyValue array to a HashMap<String, String>
    fn convert_key_value_array(array: &topt_KeyValueArray) -> HashMap<String, String> {
        let mut map = HashMap::new();
        if !array.data.is_null() {
            for i in 0..array.len {
                let pair = unsafe { &*array.data.add(i) };
                let key = if pair.key.is_null() {
                    String::new()
                } else {
                    unsafe { CStr::from_ptr(pair.key).to_string_lossy().into_owned() }
                };
                let value = if pair.value.is_null() {
                    String::new()
                } else {
                    unsafe { CStr::from_ptr(pair.value).to_string_lossy().into_owned() }
                };
                map.insert(key, value);
            }
        }
        map
    }

    /// Converts a C KeyNumber array to a HashMap<String, f64>
    fn convert_key_number_array(array: &topt_KeyNumberArray) -> HashMap<String, f64> {
        let mut map = HashMap::new();
        if !array.data.is_null() {
            for i in 0..array.len {
                let pair = unsafe { &*array.data.add(i) };
                let key = if pair.key.is_null() {
                    String::new()
                } else {
                    unsafe { CStr::from_ptr(pair.key).to_string_lossy().into_owned() }
                };
                map.insert(key, pair.value);
            }
        }
        map
    }

    /// Converts a single C topt_MockSpan to our Rust MockSpan struct
    fn convert_mock_span(mock: &topt_MockSpan) -> MockSpan {
        MockSpan {
            span_id: mock.span_id,
            trace_id: mock.trace_id,
            parent_span_id: mock.parent_span_id,
            start_time: Self::convert_unix_time(&mock.start_time),
            finish_time: Self::convert_unix_time(&mock.finish_time),
            operation_name: if mock.operation_name.is_null() {
                String::new()
            } else {
                unsafe { CStr::from_ptr(mock.operation_name).to_string_lossy().into_owned() }
            },
            string_tags: Self::convert_key_value_array(&mock.string_tags),
            number_tags: Self::convert_key_number_array(&mock.number_tags),
        }
    }

    /// Converts a C array of spans into a Vec<MockSpan>
    fn convert_mock_span_array(array: &topt_MockSpanArray) -> Vec<MockSpan> {
        let mut vec = Vec::with_capacity(array.len);
        if !array.data.is_null() {
            for i in 0..array.len {
                let mock_span = unsafe { &*array.data.add(i) };
                vec.push(Self::convert_mock_span(mock_span));
            }
        }
        vec
    }
}

File: sdks/python/test-optimization-sdk/src/test_optimization_sdk/mock_tracer.py
"""
Mock tracer module for testing.

This module provides a mock tracer implementation for testing purposes.
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Dict, List, Optional, Union

from .lib import ffi, lib


@dataclass
class MockSpan:
    """Represents a mock span for testing."""

    span_id: int
    trace_id: int
    parent_span_id: int
    start_time: datetime
    finish_time: datetime
    operation_name: str
    string_tags: Dict[str, str]
    number_tags: Dict[str, float]

    @staticmethod
    def _convert_unix_time(ut: "ffi.CData") -> datetime:
        """Convert a Unix time struct to a datetime object.

        Args:
            ut: A topt_UnixTime struct

        Returns:
            A datetime object
        """
        return datetime.fromtimestamp(ut.sec + ut.nsec / 1_000_000_000)

    @classmethod
    def from_c_struct(cls, span: "ffi.CData") -> "MockSpan":
        """Create a MockSpan from a C struct.

        Args:
            span: A topt_MockSpan struct

        Returns:
            A MockSpan instance
        """
        # Convert string tags
        string_tags: Dict[str, str] = {}
        for i in range(span.string_tags.len):
            pair = span.string_tags.data[i]
            key = ffi.string(pair.key).decode()
            value = ffi.string(pair.value).decode()
            string_tags[key] = value

        # Convert number tags
        number_tags: Dict[str, float] = {}
        for i in range(span.number_tags.len):
            pair = span.number_tags.data[i]
            key = ffi.string(pair.key).decode()
            number_tags[key] = pair.value

        return cls(
            span_id=span.span_id,
            trace_id=span.trace_id,
            parent_span_id=span.parent_span_id,
            start_time=cls._convert_unix_time(span.start_time),
            finish_time=cls._convert_unix_time(span.finish_time),
            operation_name=ffi.string(span.operation_name).decode(),
            string_tags=string_tags,
            number_tags=number_tags,
        )


class MockTracer:
    """Mock tracer for testing."""

    def __init__(self, session_id: int):
        """Initialize the mock tracer.

        Args:
            session_id: The session ID
        """
        self.session_id = session_id

    @staticmethod
    def reset() -> bool:
        """Reset the mock tracer.

        Returns:
            True if successful, False otherwise
        """
        return bool(lib.topt_debug_mock_tracer_reset())

    @staticmethod
    def get_finished_spans() -> List[MockSpan]:
        """Get all finished spans.

        Returns:
            A list of finished spans
        """
        # Get the spans
        spans = lib.topt_debug_mock_tracer_get_finished_spans()

        # Convert to Python objects
        result = []
        for i in range(spans.len):
            span = spans.data[i]
            result.append(MockSpan.from_c_struct(span))

        # Free the array
        lib.topt_debug_mock_tracer_free_mock_span_array(spans)

        return result

    @staticmethod
    def get_open_spans() -> List[MockSpan]:
        """Get all open spans.

        Returns:
            A list of open spans
        """
        # Get the spans
        spans = lib.topt_debug_mock_tracer_get_open_spans()

        # Convert to Python objects
        result = []
        for i in range(spans.len):
            span = spans.data[i]
            result.append(MockSpan.from_c_struct(span))

        # Free the array
        lib.topt_debug_mock_tracer_free_mock_span_array(spans)

        return result 